src/electronic-ids/TLV.hpp

```diff
@@ -57,10 +57,12 @@ struct TLV
 
         tag = *begin++;
         if ((tag & 0x1F) == 0x1F) { // Multi-byte tag
-            if (!*this) {
-                THROW(std::invalid_argument, "Invalid TLV: Unexpected end of tag");
-            }
-            tag = (tag << 8) | (*begin++);
+            do {
+                if (!*this) {
+                    THROW(std::invalid_argument, "Invalid TLV: Unexpected end of tag");
+                }
+                tag = (tag << 8) | (*begin++);
+            } while ((tag & 0x80) != 0x00);
```

Comment:
You have an unbounded loop here, after more than four continuation bytes the 32-bit accumulator overflows and the tag number gets trashed. I'd suggest to add a counter and check that throws once the capacity of `uint32_t` is exceeded.

Here's a unit test that proves that demonstrates the problem:

```c++
TEST(TLV, trashTag)
{
    byte_vector data {
        0x9F,  // 0x1F -> long-form
        0xF1, 0xF2, 0xF3, 0xF4, 0xF5, 0xF6, 0xF7, 0xF8, // tag bytes
        0x7F, // terminator
        0x01, // length = 1
        0x00 // dummy value byte
    };

    TLV tlv(data);
    ASSERT_TRUE(tlv);
    // This should not work, there should be an error instead.
    EXPECT_EQ(0xF6F7F87F, tlv.tag); // last three tag bytes + terminator remain
}
```

Reply:
Fixed